src/ModelContextProtocol.Core/Server/DestinationBoundMcpServer.cs

```diff
@@ -22,15 +22,25 @@ internal sealed class DestinationBoundMcpServer(McpServer server, ITransport? tr
 
     public Task SendMessageAsync(JsonRpcMessage message, CancellationToken cancellationToken = default)
     {
-        Debug.Assert(message.RelatedTransport is null);
-        message.RelatedTransport = transport;
+        if (message.Context is not null)
+        {
+            throw new ArgumentException("Only transports can provide a JsonRpcMessageContext.");
```

Comment:
would be better to not need this. If we could change the types to avoid this check, that would be best. Perhaps making the setter only accessible via an internal cast.